pandas/core/algorithms.py

```diff
@@ -1558,6 +1559,81 @@ def take_nd(arr, indexer, axis=0, out=None, fill_value=np.nan, mask_info=None,
 take_1d = take_nd
 
 
+def take_ea(arr, indexer, fill_value=_default_fill_value):
+    """Extension-array compatible take.
+
+    Parameters
+    ----------
+    arr : array-like
+        Must satisify NumPy's take semantics.
+    indexer : sequence of integers
+        Indices to be taken. See Notes for how negative indicies
+        are handled.
+    fill_value : any, optional
+        Fill value to use for NA-indicies. This has a few behaviors.
+
+        * fill_value is not specified : triggers NumPy's semantics
```

Comment:
Although, there are two things going on here:

1. `np.nan` being swapped for `ExtensionArray.dtype.na_value`.
2. `fill_value` being overloaded to switch the meaning of negative indexers.

Which bit is confusing? If I'm able to change things so that `fill_value` is simply passed through, would that be OK? Then someone specifying `ExtensionArray.take(indexer, fill_value=thing)` would mean `-1` is an NA marker. `ExtensionArray.take(indexer)` would be `-1` is the last element.

Reply:
Not overloading `fill_value` is how it is currently, with both `allow_fill` and `fill_value`. If we find that clearer, I am fine with going back to that (although it would be nice to only need one keyword for this).

I have more an objection on the np.nan swapping. It would be nice if we could prevent this, but didn't look in detail enough to assess whether this would be possible or not.

Reply:
Although, if we make sure to not pass through a default `fill_value`, you would in most cases only need to specify `allow_fill=True` to get pandas semantics, instead of needing to specify both `allow_fill` and `fill_value`. So maybe that is not that bad. 

Currenlty, eg `Series.reindex` calls under the hood (via Block.take_nd) `take_1d` on the Block values, by passing through the `Block.fill_value` as default `fill_value`. Could we make sure to not pass anything through in case of ExtensionBlocks, or to let the ExtensionBlock.fill_value be the appropriate value? (to avoid the swapping of `np.nan`)